File: samples/msgext-northwind-inventory-action-ts/src/messageExtensions/addSupplierCommand.ts
import { CardFactory, MessagingExtensionAction, MessagingExtensionActionResponse, TurnContext } from "botbuilder";
import config from '../config';
import { createSupplier } from "../northwindDB/suppliers";
import { Supplier } from "../northwindDB/model";
const COMMAND_ID = "addSupplier";
async function handleTeamsMessagingExtensionFetchTask(
    context: TurnContext,
    action: MessagingExtensionAction
): Promise<MessagingExtensionActionResponse> {
    try {
        if (action.commandId === COMMAND_ID) {
            let initialParameters = {};
            if (action.data && action.data.taskParameters) {
                initialParameters = action.data.taskParameters;
            }else {
                initialParameters = action.data
            }
            const url = `${config.botEndPoint}/public/supplier.html?p=${encodeURIComponent(JSON.stringify(initialParameters))}&appId=${config.teamsAppId}`;
            try {
                return {
                    task: {
                        type: 'continue',
                        value: {
                            width: 400,
                            height: 400,
                            title: "Add supplier",
                            url: url,
                            fallbackUrl: url
                        }
                    },
                };
            } catch (e) {
                console.error(e);
            }
        }

    } catch (e) {
        console.log(e);
    }
}
async function handleTeamsMessagingExtensionSubmitAction(
    context: TurnContext,
    action: MessagingExtensionAction
): Promise<MessagingExtensionActionResponse> {
    try {
       
        if (action.commandId === COMMAND_ID) {
            //for Copilot action
            let initialParameters = {};
            if (action.data && action.data.taskParameters) {
                initialParameters = action.data.taskParameters;
            } else {
                initialParameters = action.data
            }
            const supplier:Supplier={
                etag: "",
                partitionKey: "",
                rowKey: "",
                timestamp: new Date(),
                SupplierID: "",
                CompanyName:initialParameters["companyName"],
                ContactName:initialParameters["contactName"],
                ContactTitle:initialParameters["contactTitle"]?initialParameters["contactTitle"]:"",
                Address:initialParameters["address"]?initialParameters["address"]:"",
                City:initialParameters["city"]?initialParameters["city"]:"",
                Region:"",
                PostalCode:"",
                Country:"",
                Phone:"",
                Fax:"",
                HomePage:""
            }
            await createSupplier(supplier);            

            const heroCard = CardFactory.heroCard('Supplier added successfully', initialParameters["companyName"]);
            const attachment = {
                contentType: heroCard.contentType,
                content: heroCard.content,
                preview: heroCard
            };

            return {
                composeExtension: {
                    type: 'result',
                    attachmentLayout: 'list',
                    attachments: [
                        attachment
                    ]
                }
            };
        }

    } catch (e) {
        console.log(e);
    }
}

export default { COMMAND_ID, handleTeamsMessagingExtensionFetchTask, handleTeamsMessagingExtensionSubmitAction }

File: samples/msgext-northwind-inventory-action-csharp/MessageExtensions/AddSupplierCommand.cs
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Schema.Teams;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NorthwindInventory.NorthwindDB;
using NorthwindInventory.Models;

namespace NorthwindInventory.MessageExtensions
{
    public static class AddSupplierCommand
    {
        public const string CommandId = "addSupplier";

        public static async Task<MessagingExtensionActionResponse> HandleTeamsMessagingExtensionFetchTaskAsync(
            ITurnContext<IInvokeActivity> turnContext,
            MessagingExtensionAction action,
            IConfiguration configuration,
            CancellationToken cancellationToken)
        {
            try
            {
                if (action.CommandId == CommandId)
                {
                    var initialParameters = new Dictionary<string, object>();

                    if (action.Data is JObject dataObject && dataObject.TryGetValue("taskParameters", out var taskParameters))
                    {
                        initialParameters = taskParameters.ToObject<Dictionary<string, object>>();
                    }

                    var url = $"{configuration["BotEndPoint"]}/ClientPages/supplier.html?p={Uri.EscapeDataString(JsonConvert.SerializeObject(initialParameters))}";

                    return new MessagingExtensionActionResponse()
                    {
                        Task = new TaskModuleContinueResponse()
                        {
                            Value = new TaskModuleTaskInfo()
                            {
                                Width = 400,
                                Height = 400,
                                Title = "Add Supplier",
                                Url = url,
                                FallbackUrl = url
                            }
                        }
                    };
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
            }

            return null;
        }

        public static async Task<MessagingExtensionActionResponse> HandleTeamsMessagingExtensionSubmitActionAsync(
            ITurnContext<IInvokeActivity> turnContext,
            MessagingExtensionAction action,
            IConfiguration configuration,
            CancellationToken cancellationToken)
        {
            try
            {
                if (action.CommandId == CommandId)
                {
                    var initialParameters = new Dictionary<string, object>();

                    if (action.Data is JObject dataObject && dataObject.TryGetValue("taskParameters", out var taskParameters))
                    {
                        initialParameters = taskParameters.ToObject<Dictionary<string, object>>();
                    }
                    else if (action.Data is JObject data)
                    {
                        initialParameters = data.ToObject<Dictionary<string, object>>();
                    }

                    var supplier = new Supplier
                    {
                        ETag = string.Empty,
                        PartitionKey = string.Empty,
                        RowKey = string.Empty,
                        Timestamp = DateTime.UtcNow,
                        SupplierID = null,
                        CompanyName = initialParameters.ContainsKey("companyName") ? initialParameters["companyName"]?.ToString() : string.Empty,
                        ContactName = initialParameters.ContainsKey("contactName") ? initialParameters["contactName"]?.ToString() : string.Empty,
                        ContactTitle = string.Empty,
                        Address = string.Empty,
                        City = string.Empty,
                        Region = string.Empty,
                        PostalCode = string.Empty,
                        Country = string.Empty,
                        Phone = string.Empty,
                        Fax = string.Empty,
                        HomePage = string.Empty
                    };

                    await SupplierService.CreateSupplierAsync(supplier, configuration);

                    var heroCard = new HeroCard
                    {
                        Title = "Supplier added successfully",
                        Subtitle = initialParameters.ContainsKey("companyName") ? initialParameters["companyName"]?.ToString() : string.Empty
                    };

                    var attachment = new MessagingExtensionAttachment
                    {
                        ContentType = HeroCard.ContentType,
                        Content = heroCard,
                        Preview = heroCard.ToAttachment()
                    };

                    return new MessagingExtensionActionResponse
                    {
                        ComposeExtension = new MessagingExtensionResult
                        {
                            Type = "result",
                            AttachmentLayout = "list",
                            Attachments = new List<MessagingExtensionAttachment> { attachment }
                        }
                    };
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
            }

            return null;
        }
    }
}
